vault/identity_store_storage_util.go

```diff
@@ -0,0 +1,412 @@
+	decompressedBucketEntryBytes, err := snappy.Decode(nil, storageEntry.Value)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decompress packed storage entry: %v", err)
+	}
+
+	// JSON decode it
+	bucketEntry := &storageBucketEntry{}
+	err = jsonutil.DecodeJSON(decompressedBucketEntryBytes, bucketEntry)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode packed storage entry: %v", err)
+	}
+
+	return bucketEntry, nil
+}
+
+// upsert either inserts a new entry to the bucket or updates an existing one
+// if an entry with a matching key is already present.
+func (s *storageBucketEntry) upsert(entry *entityStorageEntry) error {
+	if s == nil {
+		return fmt.Errorf("nil storage bucket entry")
+	}
+
+	if entry == nil {
+		return fmt.Errorf("nil entry")
+	}
+
+	if entry.ID == "" {
+		return fmt.Errorf("missing entity ID")
+	}
+
+	// Look for an entry with matching key and don't modify the collection
+	// while iterating
+	foundIdx := -1
+	for itemIdx, item := range s.Items {
+		if item.ID == entry.ID {
+			foundIdx = itemIdx
+			break
+		}
+	}
+
+	// If there is no match, append the entry, otherwise update it
+	if foundIdx == -1 {
+		s.Items = append(s.Items, entry)
+	} else {
+		s.Items[foundIdx] = entry
+	}
+
+	return nil
+}
+
+// BucketIndex returns the bucket key index for a given storage entry key
+func (s *storagePacker) BucketIndex(key string) uint8 {
+	s.config.HashFunc.Reset()
+	s.config.HashFunc.Write([]byte(key))
+	return uint8(s.config.HashFunc.Sum(nil)[0])
+}
+
+// BucketKey returns the bucket key for a given entity ID
+func (s *storagePacker) BucketKey(entityID string) string {
+	return strconv.Itoa(int(s.BucketIndex(entityID)))
+}
+
+// DeleteItem removes the storage entry which the given key refers to from its
+// corresponding bucket.
+func (s *storagePacker) DeleteItem(entityID string) error {
+	if entityID == "" {
+		return fmt.Errorf("empty entity ID")
+	}
+
+	// Get the bucket key
+	bucketKey := s.BucketKey(entityID)
+
+	// Prepend the view prefix
+	bucketEntryKey := s.config.ViewPrefix + bucketKey
+
+	// Read from underlying view
+	storageEntry, err := s.view.Get(bucketEntryKey)
+	if err != nil {
+		return fmt.Errorf("failed to read packed storage entry: %v", err)
+	}
+	if storageEntry == nil {
+		return nil
+	}
+
+	// Decompress the stored value
+	decompressedBucketEntryBytes, err := snappy.Decode(nil, storageEntry.Value)
+	if err != nil {
+		return fmt.Errorf("failed to decompress packed storage entry: %v", err)
+	}
+
+	// JSON decode it
+	bucketEntry := &storageBucketEntry{}
+	err = jsonutil.DecodeJSON(decompressedBucketEntryBytes, bucketEntry)
+	if err != nil {
+		return fmt.Errorf("failed to decode packed storage entry: %v", err)
+	}
+
+	// Look for a matching storage entry
+	foundIdx := -1
+	for itemIdx, item := range bucketEntry.Items {
+		if item.ID == entityID {
+			foundIdx = itemIdx
+		}
+	}
+
+	// If there is a match, remove it from the collection and persist the
+	// resulting collection
```

Comment:
There's a lot of shared logic here between the Get and upsert functions. Maybe there's no good way to combine them but may want to see if there's a logical way to pull shared logic out into a helper function.